packages/@aws-cdk/aws-dynamodb/lib/table.ts

```diff
@@ -390,56 +450,90 @@ abstract class TableBase extends Resource implements ITable {
    */
   public abstract readonly tableStreamArn?: string;
 
+  /**
+   * Optional KMS encryption key associated with this table.
+   */
+  public abstract readonly encryptionKey?: kms.IKey;
+
   protected readonly regionalArns = new Array<string>();
 
+  /**
+   * Returns an ARN that represents all objects within the table that match
+   * the key pattern specified. To represent all keys, specify ``"*"``.
+   *
+   * If you specify multiple components for keyPattern, they will be concatenated::
+   *
+   *   arnForObjects('home/', team, '/', user, '/*')
+   *
+   */
+  public arnForObjects(keyPattern: string): string {
+    return `${this.tableArn}/${keyPattern}`;
+  }
+
   /**
    * Adds an IAM policy statement associated with this table to an IAM
    * principal's policy.
    * @param grantee The principal (no-op if undefined)
-   * @param actions The set of actions to allow (i.e. "dynamodb:PutItem", "dynamodb:GetItem", ...)
-   */
-  public grant(grantee: iam.IGrantable, ...actions: string[]): iam.Grant {
-    return iam.Grant.addToPrincipal({
+   * @param tableActions The set of actions to allow (i.e. "dynamodb:PutItem", "dynamodb:GetItem", ...)
+   * @param keyActions The set of actions to allow (i.e. "kms:Encrypt", "kms:Decrypt", ...)
+   * @param objectsKeyPattern Restrict the permission to a certain key pattern (default '*')
+   */
+  public grant(grantee: iam.IGrantable, tableActions: string[], keyActions?: string[], objectsKeyPattern?: any) {
```

Comment:
I think it would be okay for this method to *not* grant anything to the `encryptionKey` (people can add grants to the key separately using `table.encryptionKey.grant*`. This is acceptable because this is a "generic" grant... The, the `grantRead` (et al) methods that have a clear purpose (that imply a key grant needs to be extended, too) should do this proactively.

This removes the problem you had with respects to breaking this API (since it now grants _strictly_ on the table itself).

The same applies to `grantStream`.